src/lualib/Descriptors.ts

```diff
@@ -0,0 +1,72 @@
+function ____descriptorIndex(this: any, key: string): void {
+    const value = rawget(this, key);
+    if (value !== null) {
+        return value;
+    }
+
+    let metatable = getmetatable(this);
+    while (metatable) {
+        const rawResult = rawget(metatable, key);
+        if (rawResult !== undefined) {
+            return rawResult;
+        }
+
+        const descriptors = rawget(metatable, "_descriptors");
+        if (descriptors) {
+            const descriptor: PropertyDescriptor = descriptors[key];
+            if (descriptor) {
+                if (descriptor.get) {
+                    return descriptor.get.call(this);
+                }
+
+                return;
+            }
+        }
+
+        metatable = getmetatable(metatable);
+    }
+}
+
+function ____descriptorNewindex(this: any, key: string, value: any): void {
+    let metatable = getmetatable(this);
+    while (metatable) {
+        const descriptors = rawget(metatable, "_descriptors");
+        if (descriptors) {
+            const descriptor: PropertyDescriptor = descriptors[key];
+            if (descriptor) {
+                if (descriptor.set) {
+                    descriptor.set.call(this, value);
+                }
+
+                return;
+            }
+        }
+
+        metatable = getmetatable(metatable);
+    }
+
+    rawset(this, key, value);
+}
+
+function __TS__SetDescriptor(this: void, metatable: any, prop: string, descriptor: PropertyDescriptor): void {
+    if (descriptor.get) metatable.__index = ____descriptorIndex;
+    if (descriptor.set) metatable.__newindex = ____descriptorNewindex;
+    if (!metatable._descriptors) metatable._descriptors = {};
+    metatable._descriptors[prop] = descriptor;
+}
+
+function __TS__ObjectDefineProperty<T extends object>(
```

Comment:
Some comments in this file might be good, starting with a link to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty for this function (including how this function differs from that in functionality)